controllers/component/component_adapter.go

```diff
@@ -0,0 +1,114 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Adapter holds the objects needed to reconcile a integration PipelineRun.
+type Adapter struct {
+	component   *applicationapiv1alpha1.Component
+	application *applicationapiv1alpha1.Application
+	loader      loader.ObjectLoader
+	logger      h.IntegrationLogger
+	client      client.Client
+	context     context.Context
+	status      status.Status
+}
+
+// NewAdapter creates and returns an Adapter instance.
+func NewAdapter(component *applicationapiv1alpha1.Component, application *applicationapiv1alpha1.Application, logger h.IntegrationLogger, loader loader.ObjectLoader, client client.Client,
+	context context.Context) *Adapter {
+	return &Adapter{
+		component:   component,
+		application: application,
+		logger:      logger,
+		loader:      loader,
+		client:      client,
+		context:     context,
+		status:      status.NewAdapter(logger.Logger, client),
+	}
+}
+
+// EnsureComponentIsCleanedUp is an operation that will ensure components
+// marked for deletion have a snapshot created without said component
+func (a *Adapter) EnsureComponentIsCleanedUp() (controller.OperationResult, error) {
+	if !HasComponentBeenDeleted(a.component) {
+		return controller.ContinueProcessing()
+	}
+
+	applicationComponents, err := a.loader.GetAllApplicationComponents(a.client, a.context, a.application)
+	if err != nil {
+		return controller.RequeueWithError(err)
+	}
+	var updatedComponents = []applicationapiv1alpha1.Component{}
+
+	for i, individualComponent := range *applicationComponents {
+		if reflect.DeepEqual(a.component, individualComponent) {
+			updatedComponents = append((*applicationComponents)[:i], (*applicationComponents)[i+1:]...)
+		}
+	}
+
+	// create new snapshot with remaining components
+	_, err = a.RefreshSnapshot(a.client, a.application, &updatedComponents)
+	if err != nil {
+		a.logger.Error(err, "Failed to create new snapshot after component deletion")
+		return controller.RequeueWithError(err)
+	}
+
+	/* a.logger.LogAuditEvent("Snapshot integration status condition marked as Invalid, an associated snapshot components deleted.",
+	existingSnapshot, h.LogActionUpdate)
+	*/
+	return controller.ContinueProcessing()
+}
+
+// RefreshSnapshot prepares the Snapshot for a given application, components and the updated component (if any).
+// In case the Snapshot can't be created, an error will be returned.
+func (a *Adapter) RefreshSnapshot(adapterClient client.Client, application *applicationapiv1alpha1.Application, remainingComponents *[]applicationapiv1alpha1.Component) (*applicationapiv1alpha1.Snapshot, error) {
+	if len(*remainingComponents) == 0 {
+		return nil, fmt.Errorf("failed to prepare snapshot due to missing valid digest in containerImage for all components of application")
+	}
+
+	snapshot, err := gitops.PrepareSnapshot(a.client, a.context, a.application, remainingComponents, nil, "", nil)
+	if err != nil {
+		return nil, err
+	}
+
+	go metrics.RegisterNewSnapshot()
```

Comment:
Before registering a new Snapshot, we'll also need to do two things:
- Add the type label so the Snapshot is identified as a `composite` one (like. e.g. [here](https://github.com/redhat-appstudio/integration-service/blob/272ccefcfc31504768a5e28780ee2fdb144b51e3/controllers/integrationpipeline/integrationpipeline_adapter.go#L345-L348))
- Create the Snapshot on the cluster (like e.g. [here](https://github.com/redhat-appstudio/integration-service/blob/272ccefcfc31504768a5e28780ee2fdb144b51e3/controllers/integrationpipeline/integrationpipeline_adapter.go#L389-L392))